src/vdaf/mastic.rs

```diff
@@ -379,32 +391,500 @@ where
+/// parameters of Mastic used for encoding.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct MasticPrepareState<F: FieldElement, const SEED_SIZE: usize> {
+    /// Includes output shares for eventual aggregation.
+    output_shares: MasticOutputShare<F>,
+    /// If SZK verification is being performed, we also store the relevant state for that operation.
+    szk_query_state: SzkQueryState<SEED_SIZE>,
+    verifier_len: Option<usize>,
+}
+
+/// Mastic prepare share.
+///
+/// Broadcast message from an aggregator between rounds of Mastic. Includes the
+/// hashed VIDPF proofs for every prefix in the aggregation parameter, and optionally
+/// the verification message for Szk.
+#[derive(Clone, Debug)]
+pub struct MasticPrepareShare<F: FieldElement, const SEED_SIZE: usize> {
+    /// Batched VIDPF proofs
+    vidpf_proofs: Seed<SEED_SIZE>,
+    /// If Szk verification of the root weight is needed, an SZK verification message.
+    szk_query_share_opt: Option<SzkQueryShare<F, SEED_SIZE>>,
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticPrepareShare<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.vidpf_proofs.encode(bytes)?;
+        match &self.szk_query_share_opt {
+            Some(query_share) => query_share.encode(bytes),
+            None => Ok(()),
+        }
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(
+            self.vidpf_proofs.encoded_len()?
+                + match &self.szk_query_share_opt {
+                    Some(query_share) => query_share.encoded_len()?,
+                    None => 0,
+                },
+        )
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareShare<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match (&prep_state.szk_query_state, prep_state.verifier_len) {
+            (Some(_), Some(verifier_len)) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::decode(bytes)?,
+                szk_query_share_opt: Some(SzkQueryShare::<F, SEED_SIZE>::decode_with_param(
+                    &(true, verifier_len),
+                    bytes,
+                )?),
+            }),
+            (None, Some(verifier_len)) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::decode(bytes)?,
+                szk_query_share_opt: Some(SzkQueryShare::<F, SEED_SIZE>::decode_with_param(
+                    &(false, verifier_len),
+                    bytes,
+                )?),
+            }),
+            (None, None) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::<SEED_SIZE>::decode(bytes)?,
+                szk_query_share_opt: None,
+            }),
+            (Some(_), None) => Err(CodecError::UnexpectedValue),
```

Comment:
I don't think this this is parsing error, since we can only match this branch if somehow our prep state is incorrect.